[FSDP][Docs] Tidy up FSDP ctor/api docs

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105848
* __->__ #105847
* #105738

- This PR rewords the `BackwardPrefetch` docs to make the tradeoffs clear in the first sentence of each with more technical details after.
- The only supported `_FSDPPolicy` is `ModuleWrapPolicy` at the time of writing this PR. We may add others in the future such as in my other PR stack. This PR removes `_FSDPPolicy` from the public docs.
- This provides some more details around `MixedPrecision` such as explaining that layer norm and batch norm accumulate in fp32.

Follow-ups:
- Why do we force batch norm modules to have FSDP applied separately? (E.g. was this because before batch norm kernels did not support fp16/bf16?) Like layer norm, this just means that the affine parameters are in fp32. Both already accumulate in fp32 even with fp16/bf16 inputs.
- Check the `param_init_fn` + `sync_module_states=True` usage.

Commit:
[FSDP][Docs] Tidy up FSDP ctor docs

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/api.py:
```diff
@@ -71,34 +71,39 @@ class ShardingStrategy(Enum):
+    - ``BACKWARD_PRE``: This enables the most overlap but increases memory
+    usage the most. This prefetches the next set of parameters *before* the
+    current set of parameters' gradient computation. This overlaps the *next
+    all-gather* and the *current gradient computation*, and at the peak, it
```
side note, we could investigate stream priority here

rohan-varma commented:
> Check the param_init_fn + sync_module_states=True usage.

let's file issue?

awgu commented:
> > Check the param_init_fn + sync_module_states=True usage.
> 
> let's file issue?

I think someone from HF filed an issue this morning. I did not dig into it deeply yet. I did not tag the issue since I did not want to link this PR to it.

Commit:
Update on "[FSDP][Docs] Tidy up FSDP ctor/api docs"


- This PR rewords the `BackwardPrefetch` docs to make the tradeoffs clear in the first sentence of each with more technical details after.
- The only supported `_FSDPPolicy` is `ModuleWrapPolicy` at the time of writing this PR. We may add others in the future such as in my other PR stack. This PR removes `_FSDPPolicy` from the public docs.
- This provides some more details around `MixedPrecision` such as explaining that layer norm and batch norm accumulate in fp32.

Follow-ups:
- Why do we force batch norm modules to have FSDP applied separately? (E.g. was this because before batch norm kernels did not support fp16/bf16?) Like layer norm, this just means that the affine parameters are in fp32. Both already accumulate in fp32 even with fp16/bf16 inputs.
- Check the `param_init_fn` + `sync_module_states=True` usage.

[ghstack-poisoned]

awgu commented:
The failing tests are the old DTensor tests that broke due to a revert having conflicts.

awgu commented:
@pytorchbot merge -i
